File: sdks/python/curb/redact.py
"""Summarise tool arguments before they leave this process.

A human approving `delete_file` needs to know WHICH file — not the access token that
happened to sit next to it in the same kwargs. Long values are truncated and keys that
look secret are replaced by a short hash, so the approval queue stays useful without the
control plane ever storing a credential.

Mirrors packages/shared/src/redact.ts; the two must stay in step.
"""
from __future__ import annotations

import hashlib
import json
import re
from typing import Any, Dict

SECRET_KEYS = re.compile(r"pass|secret|token|key|auth|credential|cookie|session", re.I)
MAX_STRING = 120
MAX_DEPTH = 4
MAX_ITEMS = 10
MAX_KEYS = 25


def digest_args(args: Any, depth: int = 0) -> Dict[str, Any]:
    value = _redact(args, depth)
    return value if isinstance(value, dict) else {"value": value}


def fingerprint(value: Any) -> str:
    """Enough to compare two values without revealing either of them."""
    text = value if isinstance(value, str) else json.dumps(value, default=str, sort_keys=True)
    return "sha256:" + hashlib.sha256(text.encode("utf-8")).hexdigest()[:12]


def _redact(value: Any, depth: int) -> Any:
    if depth > MAX_DEPTH:
        return "[too deep]"
    if value is None or isinstance(value, (int, float, bool)):
        return value
    if isinstance(value, str):
        return _truncate(value)
    if isinstance(value, (list, tuple)):
        head = [_redact(v, depth + 1) for v in list(value)[:MAX_ITEMS]]
        extra = len(value) - MAX_ITEMS
        return head + [f"…+{extra} more"] if extra > 0 else head
    if isinstance(value, dict):
        out: Dict[str, Any] = {}
        for k, v in list(value.items())[:MAX_KEYS]:
            key = str(k)
            out[key] = fingerprint(v) if SECRET_KEYS.search(key) else _redact(v, depth + 1)
        return out
    return _truncate(str(value))


def _truncate(s: str) -> str:
    return s if len(s) <= MAX_STRING else f"{s[:MAX_STRING]}… ({len(s)} char)"

File: apps/control-plane/src/redact.ts
import { createHash } from "node:crypto";

const SECRET_KEYS = /pass|secret|token|key|auth|credential|cookie|session/i;
const MAX_STRING = 120;

/**
 * Summarise tool arguments for display in the approval queue.
 * A human needs enough context to decide ("delete WHICH file?"), but we must not
 * store secrets. Long values are truncated and suspicious keys are replaced with a
 * short hash.
 */
export function digestArgs(args: unknown, depth = 0): Record<string, unknown> {
  const value = redact(args, depth);
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : { value };
}

function redact(value: unknown, depth: number): unknown {
  if (depth > 4) return "[too deep]";
  if (value === null || value === undefined) return value;
  if (typeof value === "string") return truncate(value);
  if (typeof value === "number" || typeof value === "boolean") return value;
  if (Array.isArray(value)) {
    const head = value.slice(0, 10).map((v) => redact(v, depth + 1));
    return value.length > 10 ? [...head, `…+${value.length - 10} more`] : head;
  }
  if (typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .slice(0, 25)
        .map(([k, v]) => [k, SECRET_KEYS.test(k) ? fingerprint(v) : redact(v, depth + 1)]),
    );
  }
  return String(value);
}

function truncate(s: string): string {
  return s.length <= MAX_STRING ? s : `${s.slice(0, MAX_STRING)}… (${s.length} char)`;
}

/** Enough to compare two values without revealing either of them. */
export function fingerprint(value: unknown): string {
  const s = typeof value === "string" ? value : JSON.stringify(value ?? null);
  return `sha256:${createHash("sha256").update(s).digest("hex").slice(0, 12)}`;
}
